infrastructures/infrastructure-factory/src/main/java/org/eclipse/che/api/factory/server/scm/kubernetes/KubernetesGitCredentialManager.java

```diff
@@ -0,0 +1,167 @@
+                  s ->
+                      Boolean.parseBoolean(
+                              s.getMetadata().getAnnotations().get(ANNOTATION_GIT_CREDENTIALS))
+                          && personalAccessToken
+                              .getScmProviderUrl()
+                              .equals(s.getMetadata().getAnnotations().get(ANNOTATION_SCM_URL))
+                          && personalAccessToken
+                              .getCheUserId()
+                              .equals(s.getMetadata().getAnnotations().get(ANNOTATION_CHE_USERID))
+                          && personalAccessToken
+                              .getScmUserName()
+                              .equals(
+                                  s.getMetadata().getAnnotations().get(ANNOTATION_SCM_USERNAME)))
+              .findFirst();
+
+      Secret secret =
+          existing.orElseGet(
+              () -> {
+                Map<String, String> annotations = new HashMap<>(ANNOTATIONS);
+                annotations.put(ANNOTATION_SCM_URL, personalAccessToken.getScmProviderUrl());
+                annotations.put(ANNOTATION_SCM_USERNAME, personalAccessToken.getScmUserName());
+                annotations.put(ANNOTATION_CHE_USERID, personalAccessToken.getCheUserId());
+                ObjectMeta meta =
+                    new ObjectMetaBuilder()
+                        .withName(
+                            NameGenerator.generate(
+                                String.format(NAME_PATTERN, personalAccessToken.getScmUserName()),
+                                5))
+                        .withAnnotations(annotations)
+                        .withLabels(LABELS)
+                        .build();
+                return new SecretBuilder().withMetadata(meta).build();
+              });
+      URL scmUrl = new URL(personalAccessToken.getScmProviderUrl());
+      secret.setData(
+          Map.of(
+              "credentials",
+              Base64.getEncoder()
+                  .encodeToString(
+                      format(
+                              "%s://%s:%s@%s%s",
+                              scmUrl.getProtocol(),
+                              personalAccessToken.getScmUserName(),
+                              URLEncoder.encode(personalAccessToken.getToken(), UTF_8),
+                              scmUrl.getHost(),
+                              scmUrl.getPort() != 80 && scmUrl.getPort() != -1
+                                  ? ":" + scmUrl.getPort()
+                                  : "")
+                          .getBytes())));
+      clientFactory.create().secrets().inNamespace(namespace).createOrReplace(secret);
+    } catch (InfrastructureException | MalformedURLException e) {
+      throw new ScmConfigurationPersistenceException(e.getMessage(), e);
+    }
+  }
+
+  private String getFirstNamespace()
+      throws UnsatisfiedScmPreconditionException, ScmConfigurationPersistenceException {
+    try {
+      Optional<String> namespace =
```

Comment:
Yes, this code is sort of weak point. Credentials are user-bound, and now they are put manually by admin, so we believe that that scenario can be OK only for  <single-NS-per-user > strategy, so that code will be ok.

Reply:
It would be great if this reasoning was mentioned in the code..

Reply:
added mention